types/src/json/tx.rs

```diff
@@ -0,0 +1,618 @@
+    },
+    V2 {
+        /// This field is [flattened](https://serde.rs/field-attrs.html#flatten).
+        #[serde(flatten)]
+        prefix: TransactionPrefix,
+        rct_signatures: RctSignatures,
+        rctsig_prunable: RctSigPrunable,
+    },
+}
+
+/// [`Transaction::V1::prefix`] & [`Transaction::V2::prefix`].
+#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
+pub struct TransactionPrefix {
+    pub version: u8,
+    pub unlock_time: u64,
+    pub vin: Vec<Input>,
+    pub vout: Vec<Output>,
+    pub extra: Vec<u8>,
+}
+
+impl From<transaction::Transaction> for Transaction {
+    fn from(tx: transaction::Transaction) -> Self {
+        fn map_prefix(prefix: transaction::TransactionPrefix, version: u8) -> TransactionPrefix {
+            let mut height = 0;
+
+            let vin = prefix
+                .inputs
+                .into_iter()
+                .filter_map(|input| match input {
+                    transaction::Input::ToKey {
+                        amount,
+                        key_offsets,
+                        key_image,
+                    } => {
+                        let key = Key {
+                            amount: amount.unwrap_or(0),
+                            key_offsets,
+                            k_image: HexBytes32(key_image.compress().0),
+                        };
+
+                        Some(Input { key })
+                    }
+                    transaction::Input::Gen(h) => {
+                        height = usize_to_u64(h);
+                        None
+                    }
+                })
+                .collect();
+
+            let vout = prefix
+                .outputs
+                .into_iter()
+                .map(|o| {
+                    let amount = o.amount.unwrap_or(0);
+
+                    let target = match o.view_tag {
+                        Some(view_tag) => {
+                            let tagged_key = TaggedKey {
+                                key: HexBytes32(o.key.0),
+                                view_tag: HexBytes1([view_tag]),
+                            };
+
+                            Target::TaggedKey { tagged_key }
+                        }
+                        None => Target::Key {
+                            key: HexBytes32(o.key.0),
+                        },
+                    };
+
+                    Output { amount, target }
+                })
+                .collect();
+
+            let unlock_time = match prefix.additional_timelock {
+                transaction::Timelock::None => height,
+                transaction::Timelock::Block(height_lock) => height + usize_to_u64(height_lock),
+                transaction::Timelock::Time(seconds) => {
+                    height + (seconds / usize_to_u64(monero_serai::BLOCK_TIME))
+                }
+            } + usize_to_u64(monero_serai::DEFAULT_LOCK_WINDOW);
```

Comment:
`monero_serai` calls `unlock_time` `additional_timelock` so the values should be exactly the same - like how it's done in the DB.

The way these are differentiated is `Timelock::Time` is any value above `500_000_000`

Reply:
just to be clear this means that the inner values are what is used you don't need to do any processing

Reply:
and for `None` this should be `0`